Add media settings to product section

martinamarien opened the pull request:
**Why are these changes introduced?**
This PR adds the media global settings / class to the product template and featured product sections.

Fixes #1057

**What approach did you take?**
I applied the `global-media-settings` class to the: 
* [Product] slideshow items
* [Product] Slideshow thumbnails - Decision was made to remove box-shadow from thumbnails for technical reasons
* [Featured product] Featured Media 

I added padding - left and right on the media gallery so that the box shadow does not overlap the product content and thumbnails (when shown).

**Other considerations**
The decision was made to not include global media settings on media in the `zoom` modal, so that focus remains on the media.

**Demo links**
- [Editor](https://os2-demo.myshopify.com/admin/themes/127066177558/editor?previewPath=%2Fproducts%2Ftest-3d-models-video&context=theme&category=gid%3A%2F%2Fshopify%2FOnlineStoreThemeSettingsCategory%2FMedia%3Ftheme_id%3D127066177558%26first_setting_id%3Dmedia_border_thickness)

**Testing**
- [ ] Product media gallery options Desktop: Stacked, Thumbnail, Thumbnail Carousel
- [ ] Product media gallery options Mobile: Thumbnails, no Thumbnails
- [ ] Desktop media size: Small, Medium, Large
- [ ] Featured Media
- [ ] With and without media settings 

**Checklist**
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

tyleralsbury reviewed (dismissed):
Looks good - I really like adding the `.global-media-settings` class.

If we are going to create settings-based classes like that, I wonder if changing the naming to something like `.settings-media` to shorten it and allow for additional ones like `.settings-content`, etc... to be more easily findable using a prefix.

If we want, we can decide on that later when we do a larger refactor to handle all of these at once.

Tested...

* all media settings - desktop + mobile.
* different settings for the gallery and product page.
* Featured product

ludoboludo commented on assets/base.css:
```diff
@@ -2539,7 +2539,9 @@ details-disclosure > details {
+.global-media-settings iframe,
+.global-media-settings model-viewer,
+.global-media-settings video {
   border-radius: calc(var(--media-radius) - var(--media-border-width));
```
Should it also apply on thumbnails for `img` ? 🤔 
[screenshot](https://screenshot.click/22-01-4bkus-7g88t.png)

Maybe not something to look into in this PR but the one its using as origin.

ludoboludo reviewed (commented):
Left a few comments but let me know if it's things that should be brought up in the "origin" branch instead. 

The border seem to be inset, covering the media content: [screenshot](https://screenshot.click/22-01-jjpmx-8qts8.png). 

And in Safari I noticed that the 3D model isn't receiving the border radius as it should: [screenshot](https://screenshot.click/22-01-vka7u-17muu.png)

melissaperreault commented on assets/section-main-product.css:
```diff
@@ -465,6 +463,16 @@ a.product__text {
+
+  .product--thumbnail .product__media-list,
+  .product--thumbnail_slider .product__media-list {
+    padding-bottom: var(--media-shadow-vertical-offset);
+  }
+
+  .product__media-list {
+    padding-right: var(--media-shadow-horizontal-offset);
```
I am on the fence, merchants could add a section above and not behave properly. Merchants could get away with the new settings from https://github.com/Shopify/dawn/pull/1110 to come around a good outcome too.

martinamarien replied:
Turns out the product information section give merchants the flexibility to add padding at the top. So we don't need to worry about this.

melissaperreault commented on assets/section-main-product.css:
```diff
@@ -1033,6 +1011,7 @@ a.product__text {
   cursor: pointer;
   background-color: transparent;
   border: 0.1rem solid rgba(var(--color-foreground),.04);
```
+1

martinamarien replied:
I think I need to take some time to revisit the thumbnails. Badges can look like they're floating if I remove the outer border. Alternatively, I could keep this border and remove the one directly on the images: 

<img width="316" alt="Screen Shot 2022-01-11 at 4 59 01 PM" src="https://user-images.githubusercontent.com/26778637/149027963-0f4ec5ca-70cd-432c-b643-831fbf853c28.png">

<img width="322" alt="Screen Shot 2022-01-11 at 4 59 25 PM" src="https://user-images.githubusercontent.com/26778637/149027998-16dc4290-28ae-44b7-8815-97a62bed2f25.png">

<img width="312" alt="Screen Shot 2022-01-11 at 5 03 43 PM" src="https://user-images.githubusercontent.com/26778637/149028569-b131a0a9-3226-44e2-9d34-0730e14d18f0.png">

melissaperreault replied:
It works how it shown on your screenshot @martinamarien  👍

martinamarien replied:
Great. Thanks Meli.

ludoboludo commented on assets/section-main-product.css:
```diff
@@ -900,27 +898,7 @@ a.product__text {
-  height: calc(var(--media-border-width) * 2 + 100%);
-  top: calc(var(--media-border-width) * -1);
-  left: calc(var(--media-border-width) * -1);
-  z-index: -1;
```
You're right 👍 
It also [has an impact on the border](https://screenshot.click/22-01-10zu5-nr18r.png) when the opacity isn't set 100% but I guess that's meant to happen. So not really an issue here after all 🤔  🙌

melissaperreault replied:
If it was due by transparent background, we would seethrough the shadow underneath `mlouye` file as well no?

tyleralsbury replied:
> It also has an impact on the border when the opacity isn't set 100% but I guess that's meant to happen. So not really an issue here after all 🤔 🙌

Nothing we can do about that one.

ludoboludo commented on sections/main-product.liquid:
```diff
@@ -110,7 +110,7 @@
                 {%- endcapture -%}
                 {%- assign media_index = media_index | plus: 1 -%}
-                <li id="Slide-Thumbnails-{{ section.id }}-0" class="thumbnail-list__item slider__slide{% if section.settings.hide_variants and variant_images contains featured_media.src %} thumbnail-list_item--variant{% endif %}" data-target="{{ section.id }}-{{ featured_media.id }}" data-media-position="{{ media_index }}">
+                <li id="Slide-Thumbnails-{{ section.id }}-0" class="thumbnail-list__item slider__slide{% if section.settings.hide_variants and variant_images contains featured_media.src %} thumbnail-list_item--variant{% endif %} global-media-settings" data-target="{{ section.id }}-{{ featured_media.id }}" data-media-position="{{ media_index }}">
```
I think this class here, is the reason why there is a shadow on the first thumbnail item when set to `thumbnail` for desktop. As the first item of the list is getting the `global-media-settings` which has a `:after`.

melissaperreault reviewed (commented):
Thanks Martina!
Let's pair on the outstanding questions!

Some observations:
- [ ] Have the active link border opacity change to 100% instead + 1px border increase
- Same concern about the border thickness on the [thumbnails that are making them tiny tiny](https://screenshot.click/22-01-gxisx-ke2cr.png), we might want to consider displaying less thumbnails in view as we scale the value or systematically change the border value to a lower number on smaller media elements. _- Edit: To be explored with https://github.com/Shopify/dawn/issues/1164_
- [ ] I noticed that depending on the border thickness, we see [some of the edge or other gets cut off like shown here](https://screenshot.click/22-01-hggwk-0zv2g.png) (Will descope to Sev 3 and tackle in a separate PR)
- [x] To discuss the alignment of the thumbnails when shadow is offset to the right.

martinamarien commented:
Updates have been made: 

- [x] Add Media Border settings to product media in modal on Tablet and desktop 
- [x] Fix focus states for models and videos - on product page, modal and video section 
- [x] Fixed box-shadow overlap - Stacked layout (desktop) and slideshow layout on mobile 
- [x] Fixed reviewer feedback 
* Ludo, I _think_  I fixed https://github.com/Shopify/dawn/pull/1124#discussion_r781383345, but please let me know if I misunderstood. 

** Please test focus states in safari and chrome. Plus mouse vs. keyboard. And with different border radiuses. I was running into a bunch of issues, so hopefully I resolved them all 🤞🏻

melissaperreault reviewed (dismissed):
Apart from the gradient find with video, LGTM! 🚢 🚢 🎉 Woot!

ludoboludo commented on assets/base.css:
```diff
@@ -2533,8 +2534,18 @@ details-disclosure > details {
-.global-media-settings iframe {
+.global-media-settings iframe,
+.global-media-settings model-viewer,
+.global-media-settings video {
```
Didn't we want this to not be applied in the modal?
[video](https://screenshot.click/22-01-lxzd8-d40gg.mp4)

ludoboludo reviewed (commented):
Looking good some if not most of the thing I mentioned might not need to be tackled now 🤷

martinamarien commented on assets/section-main-product.css:
```diff
@@ -465,6 +467,15 @@ a.product__text {
 }
 
 @media screen and (min-width: 750px) {
+  .product--thumbnail .product__media-list,
```
I agree it's not ideal, but I think this is something we will need to address in a future update.

melissaperreault replied:
Looking to be fixed on this https://github.com/Shopify/dawn/issues/1164

martinamarien commented:
@tyleralsbury @ludoboludo @melissaperreault 
Should be ready for a final review

**Testing**
 - [ ] Product media gallery options Desktop: Stacked, Thumbnail, Thumbnail Carousel 
 - [ ] Product media gallery options Mobile: Thumbnails, no Thumbnails
 - [ ] Desktop media size: Small, Medium, Large
 - [ ] Featured Media
 - [ ] With and without media settings
 - [ ] Media border settings should be applied to cart item, cart notification image, product thumbnails, media in product modal (tablet/desktop only)
 - [ ] Product media, cart media and video section should use gradient class to support gradient inheritance for the border color.

tyleralsbury commented:
**Testing**

 - [ ] Product media gallery options Desktop: Stacked, Thumbnail, Thumbnail Carousel 
 - [ ] Product media gallery options Mobile: Thumbnails, no Thumbnails
 - [ ] Desktop media size: Small, Medium, Large
 - [ ] Featured Media
 - [ ] With and without media settings
 - [ ] Media border settings should be applied to cart item, cart notification image, product thumbnails, media in product modal (tablet/desktop only)
 - [ ] Product media, cart media and video section should use gradient class to support gradient inheritance for the border color.

ludoboludo reviewed (approved):
Looks good to me 👌 

I noticed an small use case where if the border radius is not super big, on focus, the main product image doesn't have a radius applied to it: 
![](https://screenshot.click/22-01-o8dan-xkx74.png)

tyleralsbury reviewed (approved):
Looks good. Found an unrelated bug where featured product being present causes a JS error.  Fixed in another PR - https://github.com/Shopify/dawn/pull/1194
